dvc/repo/experiments/executor/base.py

```diff
@@ -406,16 +412,31 @@ def commit(
             ref_info = ExpRefInfo(baseline_rev, name)
             branch = str(ref_info)
             old_ref = None
-            if not force and scm.get_ref(branch):
-                if checkpoint:
-                    raise CheckpointExistsError(ref_info.name)
-                raise ExperimentExistsError(ref_info.name)
-            logger.debug("Commit to new experiment branch '%s'", branch)
+            if scm.get_ref(branch):
+                if not force:
+                    check_exists = True
+                    logger.debug(
+                        "Reuse existing experiment branch '%s'", branch
+                    )
+                else:
+                    logger.debug(
+                        "Replace existing experiment branch '%s'", branch
+                    )
+            else:
+                logger.debug("Commit to new experiment branch '%s'", branch)
 
         scm.add([], update=True)
         scm.commit(f"dvc: commit experiment {exp_hash}", no_verify=True)
         new_rev = scm.get_rev()
-        scm.set_ref(branch, new_rev, old_ref=old_ref)
+        if check_exists:
+            orig_rev = scm.get_ref(branch)
+            if scm.diff(scm.get_ref(branch), new_rev):
+                if checkpoint:
+                    raise CheckpointExistsError(ref_info.name)
+                raise ExperimentExistsError(ref_info.name)
+            new_rev = orig_rev
```

Comment:
If the new run generates an identical commit to an existing one, we should be reusing the existing commit (this logic was already happening for tempdir runs during `git fetch`, but not for workspace runs where we commit directly into the main git/dvc workspace)

Reply:
@pmrowla, is it possible to  find out before even running this, that the state is the same as before (might be problematic  with non-deterministic stages probably).

Reply:
I think it's difficult because we also include git repo/workspace modifications with experiments (and not just DVC dependencies). So while two pipeline runs might be identical (so they get hashed into matching stages with matching DVC-tracked deps/outs), there could be other changes in the repo that show up in git but not to DVC. So we have to also generate the final git commit and then see if that actually conflicts/diffs with the previous run.

If there are git differences, we can't really tell which experiment should be preferred (so we would error out and then require running with `-f/--force` to overwrite the existing one)

Reply:
And yeah, also as you noted, I'm not sure we can rely on checkpoint stages to be deterministic, since they are persist outputs and it's not necessarily guaranteed that the user's code will always generate the identical sequence of checkpoints